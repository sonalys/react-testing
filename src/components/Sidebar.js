import PropTypes from "prop-types";
import React from "react";
import "./Sidebar.css";


const Search = ({ action }) => (
  <div className="searchContainer">
    <input className="search" placeholder="Procurar" onChange={action} />
    <img src="search.png" alt="" className="searchIcon" />
  </div>
);

const SearchResult = ({ text }) => <li>{text}</li>;

export const Sidebar = ({ searchAction, results }) => (
  <div className="container">
    <Search action={searchAction} />
    <ul>
      {results.map((item, index) => (
        <SearchResult text={item.text} key={index} />
      ))}
    </ul>
  </div>
);

Sidebar.propTypes = {
  searchAction: PropTypes.func,
  results: PropTypes.array
};

Sidebar.defaultProps = {
  searchAction: () => {},
  results: ["Slot 1", "Slot 2"]
};
