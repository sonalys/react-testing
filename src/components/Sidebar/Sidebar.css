.container {
    background-color: #dadada;
    box-shadow: 2px 0px 20px 0px #ececec;

    display: grid;
    grid-template-rows: 6vh auto;
}

.searchContainer {
    position: relative;
    margin: 0 10px;
}

.search {
    background-color: #f7f7f7;
    box-shadow: 1px 7px 14px 0px #d2d2d2fa;
    margin: 10px 0;
    height: 35px;
    width: 100%;
    border-style: none;
    text-indent: 35px;
}

.searchIcon {
    width: 20px;
    position: absolute;
    top: 19px;
    left: 8px;
}

ul {
    padding: 0;
    margin: 0;
    background-color: #cacaca;
}

li {
    background-color: #bfbfbf;
    cursor: pointer;
    margin: 0.5vh 10px 0.5vh 10px;
    list-style: none;
    padding: 0 20px;
    height: 6vh;
}

li:hover, .selected {
    margin: 0.5vh 0 0.5vh 0;
    background-color: #8d8d8d;
}