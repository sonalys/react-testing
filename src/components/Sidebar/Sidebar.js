import PropTypes from "prop-types";
import React from "react";
import "./Sidebar.css";

export const Search = ({ action }) => (
  <div className="searchContainer">
    <input className="search" placeholder="Procurar" onKeyUp={action} />
    <img src="search.png" alt="" className="searchIcon" />
  </div>
);

export const SearchResult = ({ text, selected, ...props }) => (
  <li {...props} className={selected ? "selected" : ""}>
    {text}
  </li>
);

export const Sidebar = ({ actions, selected, results }) => (
  <div className="container">
    <Search action={actions.searchAction} />
    <ul>
      {results.map((item, index) => (
        <SearchResult
          text={item.text}
          key={index}
          onClick={() => actions.selectAction(index)}
          selected={selected === index}
        />
      ))}
    </ul>
  </div>
);

Sidebar.propTypes = {
  actions: PropTypes.exact({
    searchAction: PropTypes.func,
    selectAction: PropTypes.func
  }),
  results: PropTypes.arrayOf(
    PropTypes.shape({
      text: PropTypes.string.isRequired
    })
  )
};

Sidebar.defaultProps = {
  actions: {
    searchAction: () => {},
    selectAction: () => {}
  },
  results: []
};
