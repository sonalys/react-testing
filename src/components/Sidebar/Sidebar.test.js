import "@testing-library/jest-dom/extend-expect";
import { cleanup, fireEvent, render } from "@testing-library/react";
import React from "react";
import { Search, SearchResult, Sidebar } from "./Sidebar";

afterEach(cleanup);

test("Sidebar should render", () => {
  const app = render(<Sidebar results={[{ text: "foo" }]} selected={0} />);
  expect(app).toMatchSnapshot();
});

test("Sidebar should have search", () => {
  const app = render(<Sidebar />);
  const Search = app.getByPlaceholderText("Procurar");
  expect(Search).toBeInTheDocument();
});

test("Sidebar should have list", () => {
  const app = render(<Sidebar />);
  expect(app.getByRole("list")).toBeTruthy();
});

test("Sidebar SearchList should receive props", () => {
  let list = [{ text: "foo/bar" }];
  const { getByRole, rerender } = render(<Sidebar results={list} />);
  expect(getByRole("list").childElementCount).toEqual(list.length);
  list.push({ text: "bar" });
  rerender(<Sidebar results={list} />);
  expect(getByRole("list").childElementCount).toEqual(list.length);
});

test("Search should search when typing", () => {
  const mockCallback = jest.fn();
  const app = render(<Search action={mockCallback} />);

  fireEvent.keyUp(app.getByRole("textbox"), {
    key: "A",
    code: 65,
    charCode: 65
  });
  expect(mockCallback.mock.calls.length).toBe(1);
});

test("SearchResult should render", () => {
  const app = render(<SearchResult text="foo/bar" />);
  expect(app).toMatchSnapshot();
});

test("SearchResult should dispatch click event", () => {
  const mockCallback = jest.fn();
  const app = render(<SearchResult text="foo/bar" onClick={mockCallback} />);
  fireEvent(
    app.getByRole("listitem"),
    new MouseEvent("click", { bubbles: true })
  );
  expect(mockCallback.mock.calls.length).toBe(1);
});

test("SearchResult select should alter appearance", () => {
  const app = render(<SearchResult text="this is selected" selected />);
  expect(app).toMatchSnapshot();
});
