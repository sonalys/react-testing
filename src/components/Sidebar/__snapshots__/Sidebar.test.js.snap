// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResult select should alter appearance 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="selected"
      >
        this is selected
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="selected"
    >
      this is selected
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`SearchResult should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class=""
      >
        foo/bar
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class=""
    >
      foo/bar
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Sidebar should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <div
          class="searchContainer"
        >
          <input
            class="search"
            placeholder="Procurar"
          />
          <img
            alt=""
            class="searchIcon"
            src="search.png"
          />
        </div>
        <ul>
          <li
            class="selected"
          >
            foo
          </li>
        </ul>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div
        class="searchContainer"
      >
        <input
          class="search"
          placeholder="Procurar"
        />
        <img
          alt=""
          class="searchIcon"
          src="search.png"
        />
      </div>
      <ul>
        <li
          class="selected"
        >
          foo
        </li>
      </ul>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
